### ISSUE TITLE
Transaction Built Successfully But Submit Fails

### ISSUE THREAD

**Issue by thaddeusdiamond** [Hyperion-BT/Helios#4]
Thanks for your patience in answering these :-) This may again be a Lucid interaction in how it is building the tx, but it seems to be having some issue with constructing the Time Translation:

```
{"code":2,"info":"Wallet could not send the tx.","message":"\"transaction submit error ShelleyTxValidationError ShelleyBasedEraBabbage (ApplyTxError [UtxowFailure (UtxoFailure (FromAlonzoUtxoFail (UtxosFailure (CollectErrors [BadTranslation (TimeTranslationPastHorizon \\\"PastHorizon {pastHorizonCallStack = [(\\\\\\\"runQuery\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"ouroboros-consensus-0.1.0.0-3WJNO431PEsEUwRPtCeeUr\\\\\\\", srcLocModule = \\\\\\\"Ouroboros.Consensus.HardFork.History.Qry\\\\\\\", srcLocFile = \\\\\\\"src/Ouroboros/Consensus/HardFork/History/Qry.hs\\\\\\\", srcLocStartLine = 430, srcLocStartCol = 44, srcLocEndLine = 430, srcLocEndCol = 64}),(\\\\\\\"interpretQuery\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"ouroboros-consensus-0.1.0.0-3WJNO431PEsEUwRPtCeeUr\\\\\\\", srcLocModule = \\\\\\\"Ouroboros.Consensus.HardFork.History.EpochInfo\\\\\\\", srcLocFile = \\\\\\\"src/Ouroboros/Consensus/HardFork/History/Epoc"}
```

The behavior is that Nami allows me to sign the tx, but submit returns failure.  Here is the Helios contract:

```
minting voting_ballot

const POLLS_CLOSE: Time = Time::new(1666242000000) // 10/20/22 00:00:00

enum Redeemer {
  Mint
}

func main(redeemer: Redeemer, ctx: ScriptContext) -> Bool {
  redeemer.switch {
    Mint =>  {
      tx: Tx = ctx.tx;

      tx_valid: Bool = tx.time_range.is_before(POLLS_CLOSE);
      time_range_valid_msg: String = if (tx_valid) { "Valid" } else { "Invalid" };

      print(time_range_valid_msg + " time range: " + tx.time_range.serialize().show() + " (polls close at " + POLLS_CLOSE.serialize().show() + ")");
      tx_valid
    }
  }
}
```

And here is the Javascript code building the txn:

```
const SINGLE_NFT = 1n;
const POLLS_OPEN = 1664462000000;
const POLLS_CLOSE = 1666242000000;

export async function mintBallot(blockfrostKey, policyID) {
  try {
    const cardanoDApp = CardanoDAppJs.getCardanoDAppInstance();
    validate(cardanoDApp.isWalletConnected(), 'Please connect a wallet before sweeping using "Connect Wallet" button');
    const wallet = await cardanoDApp.getConnectedWallet();

    const lucid = validated(await LucidInst.getLucidInstance(blockfrostKey), 'Please validate that your wallet is on the correct network');
    lucid.selectWallet(wallet);
    const voter = await lucid.wallet.address();

    // TODO: Get name from the user
    const assetName = toHex(new TextEncoder().encode('WildTangz 2'));

    const assetId = `${lucid.utils.mintingPolicyToId(MINTING_SCRIPT)}${assetName}`;
    const mintAssets = { [assetId]: SINGLE_NFT };
    const vendAssets = { lovelace: 2n, [assetId]: SINGLE_NFT };
    const txBuilder = lucid.newTx()
                           .addSigner(voter)
                           .mintAssets(mintAssets, Data.empty())
                           .attachMintingPolicy(MINTING_SCRIPT)
                           .payToAddress(voter, vendAssets)
                           .validFrom(POLLS_OPEN)
                           .validTo(POLLS_CLOSE - 1);

    const txComplete = await txBuilder.complete();
    const txSigned = await txComplete.sign().complete();
    const txHash = await txSigned.submit();
    shortToast(`Successfully submitted ${txHash}`);
  } catch (err) {
    shortToast(JSON.stringify(err));
  }
}
```

**Comment by christianschmitz** [comment]
You are confusing two different things.

The script should be checking if the time (or time_range, which basically represents the current time as a small window) of the tx lies within bounds of the poll. Setting the tx time_range to the poll open/close times doesn't actually provide any security.

`validTo` (and `validFrom` if you really want) should be set to a small window around the current time. Transactions will be validated within a few minutes max anyway, and I don't think you can use the Cardano mempool as a kind of temporary storage of txs for far into the future. 

Also make sure that lucid requires time in milliseconds, and not the actual slotnumbers (I'm not so sure)

**Comment by christianschmitz** [comment]
What I don't get is why validTo and validFrom are not by default set to a small window around the current time. It's what tx submitters will want 99% of the time.

**Comment by elRaulito** [comment]
ValidTo and ValidFrom are in millis so it's right, make sure to be using lucid 0.6.6

I was using 0.6.4 and had the same issue

**Comment by thaddeusdiamond** [comment]
@christianschmitz I don't think I am setting the tx time_range to poll open/close right? I am determining whether the tx is valid by using the `is_before` method. Maybe I'm misunderstanding.

@elRaulito On Lucid 0.6.6 yes.  Before Lucid was throwing the issue pre-Nami which is why I am wondering why Nami is allowing me to sign the tx but it gets rejected (presumably by Blockfrost which I believe Nami uses under the hood).

**Comment by christianschmitz** [comment]
@thaddeusdiamond but I see `validTo(POLLS_CLOSE-1)` and `validFrom(POLLS_OPEN)`? (the check inside the script is correct though)

**Comment by thaddeusdiamond** [comment]
I was confused as I thought that this would define `tx.time_range` (much like `--invalid-hereafter` on the cardano-cli.  So how is that object constructed?

**Comment by christianschmitz** [comment]
I would put `validTo(<now + 2minutes>)`, and leave out validFrom which you don't really need in this example

**Comment by elRaulito** [comment]
If you wish you can paste here the cbor file, when you create the tx with lucid instead of .complete(); use .toString(); and console.log the result, so we can see the slot equivalent in the cbor and see why it's throwing error

**Comment by christianschmitz** [comment]
Txs built with Helios allow that: toCbor is defined on the `Tx` instance itself

**Comment by thaddeusdiamond** [comment]
Here's the JSON representation:

```js
{
  "body": {
    "original_bytes": null,
    "inputs": [
      {
        "transaction_id": "afcaa4fd58c039a2d8139e519a0c9b5b4ea0a80bcf71ff593edae0264c1df598",
        "index": "0"
      }
    ],
    "outputs": [
      {
        "address": "addr_test1qqkjh9xrj566jmm5em0epzmsfmdfvpw86nnc3g8f7zsn83u6u3a9lm3zqxwg6sn2m93cm0leqtam6apzje45xcw9dq5se26law",
        "amount": {
          "coin": "1168010",
          "multiasset": {
            "9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5f": {
              "57696c6454616e677a2032": "1"
            }
          }
        },
        "datum": null,
        "script_ref": null
      },
      {
        "address": "addr_test1qqkjh9xrj566jmm5em0epzmsfmdfvpw86nnc3g8f7zsn83u6u3a9lm3zqxwg6sn2m93cm0leqtam6apzje45xcw9dq5se26law",
        "amount": {
          "coin": "9998511152",
          "multiasset": null
        },
        "datum": null,
        "script_ref": null
      }
    ],
    "fee": "320838",
    "ttl": "10558801",
    "certs": null,
    "withdrawals": null,
    "update": null,
    "auxiliary_data_hash": null,
    "validity_start_interval": "8778801",
    "mint": {
      "9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5f": {
        "57696c6454616e677a2032": "1"
      }
    },
    "script_data_hash": "8061244764c5c9ee1e9d9dc39ac9a5f3d9ac7650afac7472a1cfd594baf7b08c",
    "collateral": [
      {
        "transaction_id": "afcaa4fd58c039a2d8139e519a0c9b5b4ea0a80bcf71ff593edae0264c1df598",
        "index": "0"
      }
    ],
    "required_signers": [
      "2d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c7"
    ],
    "network_id": null,
    "collateral_return": {
      "address": "addr_test1qqkjh9xrj566jmm5em0epzmsfmdfvpw86nnc3g8f7zsn83u6u3a9lm3zqxwg6sn2m93cm0leqtam6apzje45xcw9dq5se26law",
      "amount": {
        "coin": "9999518743",
        "multiasset": null
      },
      "datum": null,
      "script_ref": null
    },
    "total_collateral": "481257",
    "reference_inputs": null
  },
  "witness_set": {
    "vkeys": null,
    "native_scripts": null,
    "bootstraps": null,
    "plutus_scripts": null,
    "plutus_data": null,
    "redeemers": [
      {
        "tag": "Mint",
        "index": "0",
        "data": {
          "datum": {
            "ConstrPlutusData": {
              "alternative": "0",
              "data": {
                "elems": [],
                "definite_encoding": false
              }
            }
          },
          "original_bytes": [
            216,
            121,
            159,
            255
          ]
        },
        "ex_units": {
          "mem": "1121567",
          "steps": "652739025"
        }
      }
    ],
    "plutus_v2_scripts": [
      "59026b0100003232323232323232323232323232323232323232323232323232322233335734660060040024931324c44646644646464666ae70c088cc080cc080cc080cc080cc080004c08524010d2074696d652072616e67653a20003301e3301c30120034992630214901112028706f6c6c7320636c6f736520617420003301e3301100949926302149010129002003498cccd5cd00091810a4810556616c696400232302249107496e76616c696400499263300b3010001007300900435573a6ea8004004008c004dd424101a53379fcc0002400200844646644646666ae68cdc3800a400849408cccd5cd19b870024800092891999ab9a0042337106eb4d5d09aab9e375400c6eb40208cdc49bad357426aae78dd50031bad00849926498d55ce9baa0023006001300530040013003002235742601a0024666ae68cdc39aab9d375400290002504a246ae84c0300040300048d5d0980100091aba23002001235744600400246ae88c0080048d5d1180100091aba23002001235744600400246ae88c0080048d55cf1baa0010012237526ecc0088c88c8c01ccc00400400c88c8cccd5cd19b88480000048cdcb1b9932333573466e1cdc6800a400466e2d2060001001375c660106ea0cdc7001a40009319802002199b8c4800800800c924100498dc68009bae00123223237526666ae68cdc4001a4000466e2d205a3300200233704008900191980100100224c4464646666ae68cdc4001a40404004466e28cc014014cdc1802240400049319b8b333573466e20005201433700002903019b80001482b805221003370c00290101bad001232230043372c004600a002600600246ea4dcc00091b99375c003"
    ]
  },
  "is_valid": true,
  "auxiliary_data": null
}
```

And here's the CBOR:

```
84ab0081825820afcaa4fd58c039a2d8139e519a0c9b5b4ea0a80bcf71ff593edae0264c1df598000182825839002d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c79ae47a5fee22019c8d426ad9638dbff902fbbd7422966b4361c56829821a0011d28aa1581c9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5fa14b57696c6454616e677a203201825839002d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c79ae47a5fee22019c8d426ad9638dbff902fbbd7422966b4361c568291b0000000253f52c30021a0004e546031a00a11d51081a0085f43109a1581c9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5fa14b57696c6454616e677a2032010b58208061244764c5c9ee1e9d9dc39ac9a5f3d9ac7650afac7472a1cfd594baf7b08c0d81825820afcaa4fd58c039a2d8139e519a0c9b5b4ea0a80bcf71ff593edae0264c1df598000e81581c2d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c710825839002d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c79ae47a5fee22019c8d426ad9638dbff902fbbd7422966b4361c568291b0000000254048c17111a000757e9a20581840100d8799fff821a00111d1f1a26e801d1068159026e59026b0100003232323232323232323232323232323232323232323232323232322233335734660060040024931324c44646644646464666ae70c088cc080cc080cc080cc080cc080004c08524010d2074696d652072616e67653a20003301e3301c30120034992630214901112028706f6c6c7320636c6f736520617420003301e3301100949926302149010129002003498cccd5cd00091810a4810556616c696400232302249107496e76616c696400499263300b3010001007300900435573a6ea8004004008c004dd424101a53379fcc0002400200844646644646666ae68cdc3800a400849408cccd5cd19b870024800092891999ab9a0042337106eb4d5d09aab9e375400c6eb40208cdc49bad357426aae78dd50031bad00849926498d55ce9baa0023006001300530040013003002235742601a0024666ae68cdc39aab9d375400290002504a246ae84c0300040300048d5d0980100091aba23002001235744600400246ae88c0080048d5d1180100091aba23002001235744600400246ae88c0080048d55cf1baa0010012237526ecc0088c88c8c01ccc00400400c88c8cccd5cd19b88480000048cdcb1b9932333573466e1cdc6800a400466e2d2060001001375c660106ea0cdc7001a40009319802002199b8c4800800800c924100498dc68009bae00123223237526666ae68cdc4001a4000466e2d205a3300200233704008900191980100100224c4464646666ae68cdc4001a40404004466e28cc014014cdc1802240400049319b8b333573466e20005201433700002903019b80001482b805221003370c00290101bad001232230043372c004600a002600600246ea4dcc00091b99375c003f5f6
```

**Comment by thaddeusdiamond** [comment]
Also, changing it to `validTo(POLLS_CLOSE + 120000)` and removing `validFrom` didn't change anything.

**Comment by elRaulito** [comment]
Yes sorry, I wanted to know the slots in the cbor, maybe it was a wrong slot "translation"

**Comment by thaddeusdiamond** [comment]
Ah gotcha.  Above says ttl "10558801" because it's preprod.  But when I do the same Tx building and no minting policy the tx submits fine.

**Comment by elRaulito** [comment]
You mean, that if you build a transaction with                 .validFrom(POLLS_OPEN) .validTo(POLLS_CLOSE - 1); 
without minting it works?

**Comment by christianschmitz** [comment]
Both ttl seems too far from the current slot number. I recommend using current time (i.e. `new Date().getTime()`), and adding a few minutes. I wouldn't use the poll interval itself when building the tx

**Comment by thaddeusdiamond** [comment]
Mainnet had no impact on the submission.  Getting this error:

```
{"code":2,"info":"Wallet could not send the tx.","message":"\"transaction submit error ShelleyTxValidationError ShelleyBasedEraBabbage (ApplyTxError [UtxowFailure (UtxoFailure (FromAlonzoUtxoFail (UtxosFailure (CollectErrors [BadTranslation (TimeTranslationPastHorizon \\\"PastHorizon {pastHorizonCallStack = [(\\\\\\\"runQuery\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"ouroboros-consensus-0.1.0.0-3WJNO431PEsEUwRPtCeeUr\\\\\\\", srcLocModule = \\\\\\\"Ouroboros.Consensus.HardFork.History.Qry\\\\\\\", srcLocFile = \\\\\\\"src/Ouroboros/Consensus/HardFork/History/Qry.hs\\\\\\\", srcLocStartLine = 430, srcLocStartCol = 44, srcLocEndLine = 430, srcLocEndCol = 64}),(\\\\\\\"interpretQuery\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"ouroboros-consensus-0.1.0.0-3WJNO431PEsEUwRPtCeeUr\\\\\\\", srcLocModule = \\\\\\\"Ouroboros.Consensus.HardFork.History.EpochInfo\\\\\\\", srcLocFile = \\\\\\\"src/Ouroboros/Consensus/HardFork/History/EpochInfo.hs\\\\\\\", srcLocStartLine = 45, srcLocStartCol = 50, srcLocEndLine = 45, srcLocEndCol = 68}),(\\\\\\\"interpretQuery'\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"ouroboros-consensus-0.1.0.0-3WJNO431PEsEUwRPtCeeUr\\\\\\\", srcLocModule = \\\\\\\"Ouroboros.Consensus.HardFork.History.EpochInfo\\\\\\\", srcLocFile = \\\\\\\"src/Ouroboros/Consensus/HardFork/History/EpochInfo.hs\\\\\\\", srcLocStartLine = 39, srcLocStartCol = 9, srcLocEndLine = 39, srcLocEndCol = 52}),(\\\\\\\"epochInfoSlotToRelativeTime_\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"cardano-slotting-0.1.0.0-6VyoKllhKMr9HEkTJGSq3p\\\\\\\", srcLocModule = \\\\\\\"Cardano.Slotting.EpochInfo.API\\\\\\\", srcLocFile = \\\\\\\"src/Cardano/Slotting/EpochInfo/API.hs\\\\\\\", srcLocStartLine = 62, srcLocStartCol = 9, srcLocEndLine = 62, srcLocEndCol = 37}),(\\\\\\\"epochInfoSlotToRelativeTime\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"cardano-slotting-0.1.0.0-6VyoKllhKMr9HEkTJGSq3p\\\\\\\", srcLocModule = \\\\\\\"Cardano.Slotting.EpochInfo.API\\\\\\\", srcLocFile = \\\\\\\"src/Cardano/Slotting/EpochInfo/API.hs\\\\\\\", srcLocStartLine = 125, srcLocStartCol = 40, srcLocEndLine = 125, srcLocEndCol = 70}),(\\\\\\\"epochInfoSlotToRelativeTime_\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"cardano-slotting-0.1.0.0-6VyoKllhKMr9HEkTJGSq3p\\\\\\\", srcLocModule = \\\\\\\"Cardano.Slotting.EpochInfo.API\\\\\\\", srcLocFile = \\\\\\\"src/Cardano/Slotting/EpochInfo/API.hs\\\\\\\", srcLocStartLine = 62, srcLocStartCol = 9, srcLocEndLine = 62, srcLocEndCol = 37}),(\\\\\\\"epochInfoSlotToRelativeTime\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"cardano-slotting-0.1.0.0-6VyoKllhKMr9HEkTJGSq3p\\\\\\\", srcLocModule = \\\\\\\"Cardano.Slotting.EpochInfo.API\\\\\\\", srcLocFile = \\\\\\\"src/Cardano/Slotting/EpochInfo/API.hs\\\\\\\", srcLocStartLine = 91, srcLocStartCol = 30, srcLocEndLine = 91, srcLocEndCol = 63}),(\\\\\\\"epochInfoSlotToUTCTime\\\\\\\",SrcLoc {srcLocPackage = \\\\\\\"cardano-ledger-alonzo-0.1.0.0-DgdteOlr4HbEVKVu4Mugad\\\\\\\", srcLocModule = \\\\\\\"Cardano.Ledger.Alonzo.TxInfo\\\\\\\", srcLocFile = \\\\\\\"src/Cardano/Ledger/Alonzo/TxInfo.hs\\\\\\\", srcLocStartLine = 233, srcLocStartCol = 9, srcLocEndLine = 233, srcLocEndCol = 41})], pastHorizonExpression = Some (EPair (ERelToAbsTime (ERelSlotToTime (EAbsToRelSlot (ELit (SlotNo 74675829))))) (ESlotLength (ELit (SlotNo 74675829)))), pastHorizonSummary = [EraSummary {eraStart = Bound {boundTime = RelativeTime 0s, boundSlot = SlotNo 0, boundEpoch = EpochNo 0}, eraEnd = EraEnd (Bound {boundTime = RelativeTime 89856000s, boundSlot = SlotNo 4492800, boundEpoch = EpochNo 208}), eraParams = EraParams {eraEpochSize = EpochSize 21600, eraSlotLength = SlotLength 20s, eraSafeZone = StandardSafeZone 4320}},EraSummary {eraStart = Bound {boundTime = RelativeTime 89856000s, boundSlot = SlotNo 4492800, boundEpoch = EpochNo 208}, eraEnd = EraEnd (Bound {boundTime = RelativeTime 101952000s, boundSlot = SlotNo 16588800, boundEpoch = EpochNo 236}), eraParams = EraParams {eraEpochSize = EpochSize 432000, eraSlotLength = SlotLength 1s, eraSafeZone = StandardSafeZone 129600}},EraSummary {eraStart = Bound {boundTime = RelativeTime 101952000s, boundSlot = SlotNo 16588800, boundEpoch = EpochNo 236}, eraEnd = EraEnd (Bound {boundTime = RelativeTime 108432000s, boundSlot = SlotNo 23068800, boundEpoch = EpochNo 251}), eraParams = EraParams {eraEpochSize = EpochSize 432000, eraSlotLength = SlotLength 1s, eraSafeZone = StandardSafeZone 129600}},EraSummary {eraStart = Bound {boundTime = RelativeTime 108432000s, boundSlot = SlotNo 23068800, boundEpoch = EpochNo 251}, eraEnd = EraEnd (Bound {boundTime = RelativeTime 125280000s, boundSlot = SlotNo 39916800, boundEpoch = EpochNo 290}), eraParams = EraParams {eraEpochSize = EpochSize 432000, eraSlotLength = SlotLength 1s, eraSafeZone = StandardSafeZone 129600}},EraSummary {eraStart = Bound {boundTime = RelativeTime 125280000s, boundSlot = SlotNo 39916800, boundEpoch = EpochNo 290}, eraEnd = EraEnd (Bound {boundTime = RelativeTime 157680000s, boundSlot = SlotNo 72316800, boundEpoch = EpochNo 365}), eraParams = EraParams {eraEpochSize = EpochSize 432000, eraSlotLength = SlotLength 1s, eraSafeZone = StandardSafeZone 129600}},EraSummary {eraStart = Bound {boundTime = RelativeTime 157680000s, boundSlot = SlotNo 72316800, boundEpoch = EpochNo 365}, eraEnd = EraEnd (Bound {boundTime = RelativeTime 158544000s, boundSlot = SlotNo 73180800, boundEpoch = EpochNo 367}), eraParams = EraParams {eraEpochSize = EpochSize 432000, eraSlotLength = SlotLength 1s, eraSafeZone = StandardSafeZone 129600}}]}\\\")])))),UtxowFailure (UtxoFailure (FromAlonzoUtxoFail (OutsideForecast (SlotNo 74675829))))])\""}
```

Here is the tx in JSON:

```
{
  "body": {
    "original_bytes": null,
    "inputs": [
      {
        "transaction_id": "b5b68ee8abeb09670dd7884e6482577327b3efbff3c624fb3a5887901d530ca7",
        "index": "2"
      }
    ],
    "outputs": [
      {
        "address": "addr1qykjh9xrj566jmm5em0epzmsfmdfvpw86nnc3g8f7zsn83u6u3a9lm3zqxwg6sn2m93cm0leqtam6apzje45xcw9dq5s6u8l33",
        "amount": {
          "coin": "1168010",
          "multiasset": {
            "9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5f": {
              "57696c6454616e677a2032": "1"
            }
          }
        },
        "datum": null,
        "script_ref": null
      },
      {
        "address": "addr1qykjh9xrj566jmm5em0epzmsfmdfvpw86nnc3g8f7zsn83u6u3a9lm3zqxwg6sn2m93cm0leqtam6apzje45xcw9dq5s6u8l33",
        "amount": {
          "coin": "29559906",
          "multiasset": null
        },
        "datum": null,
        "script_ref": null
      }
    ],
    "fee": "301755",
    "ttl": "74675829",
    "certs": null,
    "withdrawals": null,
    "update": null,
    "auxiliary_data_hash": null,
    "validity_start_interval": null,
    "mint": {
      "9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5f": {
        "57696c6454616e677a2032": "1"
      }
    },
    "script_data_hash": "6d307869f1464f7ba56b3e4c39c12d01b67e3b396c20b920c415c071add3d9fd",
    "collateral": [
      {
        "transaction_id": "b5b68ee8abeb09670dd7884e6482577327b3efbff3c624fb3a5887901d530ca7",
        "index": "2"
      }
    ],
    "required_signers": [
      "2d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c7"
    ],
    "network_id": null,
    "collateral_return": {
      "address": "addr1qykjh9xrj566jmm5em0epzmsfmdfvpw86nnc3g8f7zsn83u6u3a9lm3zqxwg6sn2m93cm0leqtam6apzje45xcw9dq5s6u8l33",
      "amount": {
        "coin": "30577038",
        "multiasset": null
      },
      "datum": null,
      "script_ref": null
    },
    "total_collateral": "452633",
    "reference_inputs": null
  },
  "witness_set": {
    "vkeys": null,
    "native_scripts": null,
    "bootstraps": null,
    "plutus_scripts": null,
    "plutus_data": null,
    "redeemers": [
      {
        "tag": "Mint",
        "index": "0",
        "data": {
          "datum": {
            "ConstrPlutusData": {
              "alternative": "0",
              "data": {
                "elems": [],
                "definite_encoding": false
              }
            }
          },
          "original_bytes": [
            216,
            121,
            159,
            255
          ]
        },
        "ex_units": {
          "mem": "942862",
          "steps": "539610849"
        }
      }
    ],
    "plutus_v2_scripts": [
      "59026b0100003232323232323232323232323232323232323232323232323232322233335734660060040024931324c44646644646464666ae70c088cc080cc080cc080cc080cc080004c08524010d2074696d652072616e67653a20003301e3301c30120034992630214901112028706f6c6c7320636c6f736520617420003301e3301100949926302149010129002003498cccd5cd00091810a4810556616c696400232302249107496e76616c696400499263300b3010001007300900435573a6ea8004004008c004dd424101a53379fcc0002400200844646644646666ae68cdc3800a400849408cccd5cd19b870024800092891999ab9a0042337106eb4d5d09aab9e375400c6eb40208cdc49bad357426aae78dd50031bad00849926498d55ce9baa0023006001300530040013003002235742601a0024666ae68cdc39aab9d375400290002504a246ae84c0300040300048d5d0980100091aba23002001235744600400246ae88c0080048d5d1180100091aba23002001235744600400246ae88c0080048d55cf1baa0010012237526ecc0088c88c8c01ccc00400400c88c8cccd5cd19b88480000048cdcb1b9932333573466e1cdc6800a400466e2d2060001001375c660106ea0cdc7001a40009319802002199b8c4800800800c924100498dc68009bae00123223237526666ae68cdc4001a4000466e2d205a3300200233704008900191980100100224c4464646666ae68cdc4001a40404004466e28cc014014cdc1802240400049319b8b333573466e20005201433700002903019b80001482b805221003370c00290101bad001232230043372c004600a002600600246ea4dcc00091b99375c003"
    ]
  },
  "is_valid": true,
  "auxiliary_data": null
}
```

And here is the CBOR:

```
84aa0081825820b5b68ee8abeb09670dd7884e6482577327b3efbff3c624fb3a5887901d530ca7020182825839012d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c79ae47a5fee22019c8d426ad9638dbff902fbbd7422966b4361c56829821a0011d28aa1581c9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5fa14b57696c6454616e677a203201825839012d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c79ae47a5fee22019c8d426ad9638dbff902fbbd7422966b4361c568291a01c30c62021a00049abb031a0473767509a1581c9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5fa14b57696c6454616e677a2032010b58206d307869f1464f7ba56b3e4c39c12d01b67e3b396c20b920c415c071add3d9fd0d81825820b5b68ee8abeb09670dd7884e6482577327b3efbff3c624fb3a5887901d530ca7020e81581c2d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c710825839012d2b94c39535a96f74cedf908b704eda9605c7d4e788a0e9f0a133c79ae47a5fee22019c8d426ad9638dbff902fbbd7422966b4361c568291a01d2918e111a0006e819a20581840100d8799fff821a000e630e1a2029cee1068159026e59026b0100003232323232323232323232323232323232323232323232323232322233335734660060040024931324c44646644646464666ae70c088cc080cc080cc080cc080cc080004c08524010d2074696d652072616e67653a20003301e3301c30120034992630214901112028706f6c6c7320636c6f736520617420003301e3301100949926302149010129002003498cccd5cd00091810a4810556616c696400232302249107496e76616c696400499263300b3010001007300900435573a6ea8004004008c004dd424101a53379fcc0002400200844646644646666ae68cdc3800a400849408cccd5cd19b870024800092891999ab9a0042337106eb4d5d09aab9e375400c6eb40208cdc49bad357426aae78dd50031bad00849926498d55ce9baa0023006001300530040013003002235742601a0024666ae68cdc39aab9d375400290002504a246ae84c0300040300048d5d0980100091aba23002001235744600400246ae88c0080048d5d1180100091aba23002001235744600400246ae88c0080048d55cf1baa0010012237526ecc0088c88c8c01ccc00400400c88c8cccd5cd19b88480000048cdcb1b9932333573466e1cdc6800a400466e2d2060001001375c660106ea0cdc7001a40009319802002199b8c4800800800c924100498dc68009bae00123223237526666ae68cdc4001a4000466e2d205a3300200233704008900191980100100224c4464646666ae68cdc4001a40404004466e28cc014014cdc1802240400049319b8b333573466e20005201433700002903019b80001482b805221003370c00290101bad001232230043372c004600a002600600246ea4dcc00091b99375c003f5f6
```

So let's say I change the date using `validTo(new Date().getTime() + <5 mins>)`.  Now I am getting this error:

```
{"code":2,"info":"Wallet could not send the tx.","message":"\"transaction submit error ShelleyTxValidationError ShelleyBasedEraBabbage (ApplyTxError [UtxowFailure (UtxoFailure (FromAlonzoUtxoFail (UtxosFailure (ValidationTagMismatch (IsValid True) (FailedUnexpectedly (PlutusFailure \\\"\\\\nThe 2 arg plutus script (PlutusScript PlutusV2 ScriptHash \\\\\\\"9ad039bea25f377016faa140fbdc1c5af3f27ffa5b74f8235da1ce5f\\\\\\\") fails.\\\\nCekError An error has occurred: User error:\\\\nThe machine terminated part way through evaluation due to overspending the budget.\\\\nThe budget when the machine terminated was:\\\\n({ cpu: 4112974\\\\n| mem: -2581\\\\n})\\\\nNegative numbers indicate the overspent budget; note that this only indicatessthe budget that was needed for the next step, not to run the program to completion.\\\\nThe protocol version is: ProtVer {pvMajor = 7, pvMinor = 0}\\\\nThe re"}
```

Are you compiling with Plutonomy? I worked with JPG on similar types of overspending issues and they said Plutonomy (or Plutarch? can't remember), might help.  The helios script is below:

```
minting voting_ballot

const POLLS_CLOSE: Time = Time::new(1666242000000) // 10/20/22 00:00:00

enum Redeemer {
  Mint
}

func main(redeemer: Redeemer, ctx: ScriptContext) -> Bool {
  redeemer.switch {
    Mint =>  {
      tx: Tx = ctx.tx;

      tx_valid: Bool = tx.time_range.is_before(POLLS_CLOSE);
      time_range_valid_msg: String = if (tx_valid) { "Valid" } else { "Invalid" };

      print(time_range_valid_msg + " time range: " + tx.time_range.serialize().show() + " (polls close at " + POLLS_CLOSE.serialize().show() + ")");
      tx_valid
    }
  }
}
```

**Comment by christianschmitz** [comment]
Budget overspending error, that means Lucid (and it's backend) aren't correctly estimating the execution budget

**Comment by thaddeusdiamond** [comment]
Alright.  Let me try to use your Tx building API in Javascript and get back to you.  Can you provide that link in this ticket one more time for myself and others? Thanks for all your help.

**Comment by christianschmitz** [comment]
Examples of using helios tx building functionality can be found in following repository: https://github.com/Hyperion-BT/PicoSwap

Files of interest are `contract.js` and `index.js` (eg. `buyContract` function)

**Comment by christianschmitz** [comment]
PicoSwap in action (chrome only): https://hyperion-bt.github.io/PicoSwap/

**Comment by thaddeusdiamond** [comment]
Thank you! Will report back soon.

**Comment by rodrigomd94** [comment]
I think this is also an issue with the native uplc compiler used when building the tx with Lucid. You can try the solution proposed in issue #5

**Comment by thaddeusdiamond** [comment]
Yes! Worked like a charm when I did `const txComplete = await txBuilder.complete({ nativeUplc: false });` then it compiled fine (which means that this is perhaps an issue in the current version of Aiken? These seem to be the relevant lines for those curious: https://github.com/berry-pool/cardano-multiplatform-lib/blob/vasil/rust/src/tx_builder.rs#L2426-L2445).

Defaulting back to Blockfrost works. Closing this issue, thanks.
